### DISCUSSION TITLE
How does search query work?

### DISCUSSION THREAD

**Discussion by pitcairn1987** [appwrite/appwrite#5005] [Q&A] [answered]
### 👟 Reproduction steps

II'm designing a mobile app search filter and I have a fullindex on 2 attributes.

The Query.search method works strangely. For example, if I have the string "1111,2222", method finds the record both if I pass the string "2222" or "1111". But if the string looks like this "1111,ww" and I type "ww" it no longer finds this record. It only finds if I type "1111".

But this is not the only case where it works in unexpected ways.
For example, I have the string "11111hutat". When I type "hutat" it doesn't find me a record. But if I enter "11111 hutat", or "11111;hutat", so with some separator, it already finds.
![3](https://user-images.githubusercontent.com/5616509/212277264-4bbc390e-c4a8-49c4-bbcd-f89e00cd8696.jpg)
![2](https://user-images.githubusercontent.com/5616509/212277269-8b6f454a-c6bf-449d-8b6f-a48f67d04abf.jpg)
![1](https://user-images.githubusercontent.com/5616509/212277272-93d49a5d-b6d9-47fe-95fe-7cac4b521577.jpg)

I was wondering if it might have something to do with caching some things by backend, because It works in a completely unpredictable way.

### 👍 Expected behavior

Query.search method should always search in the whole string some substring, not from the beginning, 

### 👎 Actual Behavior

It works in a completely unpredictable way. It only works always fine if I have a string of digits separated by a separator

### 🎲 Appwrite version

Version 1.0.x

### 💻 Operating system

Linux

### 🧱 Your Environment

I use appwrite as backend for Flutte mobile app.

### 👀 Have you spent some time to check if this issue has been raised before?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Code of Conduct?

- [X] I have read the [Code of Conduct](https://github.com/appwrite/appwrite/blob/HEAD/CODE_OF_CONDUCT.md)

**Comment by stnguyen90** [comment, marked as answer]
Appwrite uses [MariaDB's full-text search](https://mariadb.com/kb/en/full-text-index-overview/) in [boolean mode](https://mariadb.com/kb/en/full-text-index-overview/#in-boolean-mode) for the search query. When you make a query like:

```javascript
Query.search("yearTags", ["1111"])
```

Appwrite sends a query to MariaDB like:

```sql
MATCH(yearTags) AGAINST("'1111*'" IN BOOLEAN MODE)
```

> **Note**: The `*` at the end of your query value does a wildcard to support auto-complete.

[source code](https://github.com/utopia-php/database/blob/2590b533654cf4ac0442ed83dc143a23b47dd61c/src/Database/Adapter/MariaDB.php#L1683-L1702)

Since we're using MariaDB's full-text search, there are a few other things to keep in mind.

1. [Only words are indexed](https://dev.mysql.com/doc/refman/8.0/en/fulltext-natural-language.html). A "word" is considered "any sequence of true word characters (letters, digits, and underscores) as a word." A word may also contain 1 apostrophe.
2. Words less than 3 characters are ignored
3. Words greater than 84 characters are ignored

I hope that gives you more insight into how the search works.

**Reply by pitcairn1987** [reply]
Thanks!! I would never have guessed it:)

**Reply by xuelink** [reply]
Are there rules are still valid ? Because i can use search query with 1 char now.
